### PR TITLE
Minor: Improve `ExecutionPlan` documentation

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

N/A

## Rationale for this change
While reviewing https://github.com/apache/arrow-datafusion/pull/8013 from @tustvold  and @comphead I noticed there were other parts of the documentation that could be improved to help people implement these interfaces. 

## What changes are included in this PR?

Update doc comments in `ExecutionPlan` to try and lower the barrier to implementing one (I think @tustvold 's examples are probably 90% of the value already)

## Are these changes tested?
Yes, by existing CI tests

## Are there any user-facing changes?

Docs

### CODE REVIEW

File: datafusion/physical-plan/src/lib.rs [hunk 1]
```diff
@@ -91,16 +91,26 @@ pub use datafusion_physical_expr::{
...
 
-/// `ExecutionPlan` represent nodes in the DataFusion Physical Plan.
+/// Represent nodes in the DataFusion Physical Plan.
 ///
-/// Each `ExecutionPlan` is partition-aware and is responsible for
-/// creating the actual `async` [`SendableRecordBatchStream`]s
-/// of [`RecordBatch`] that incrementally compute the operator's
-/// output from its input partition.
+/// Calling [`execute`] produces an `async` [`SendableRecordBatchStream`] of
+/// [`RecordBatch`] that incrementally computes a partition of the
```
Comment (comphead) [inline datafusion/physical-plan/src/lib.rs:97]: when we talk about partition we mean a separate task(thread) that processes a chunk of input data by applying `.execute()` method?

File: datafusion/physical-plan/src/lib.rs [hunk 2]
```diff
@@ -166,19 +179,21 @@ pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
...
-    /// Specifies whether the operator benefits from increased parallelization
-    /// at its input for each child. If set to `true`, this indicates that the
-    /// operator would benefit from partitioning its corresponding child
-    /// (and thus from more parallelism). For operators that do very little work
-    /// the overhead of extra parallelism may outweigh any benefits
+    /// Specifies whether the `ExecutionPlan` benefits from increased
+    /// parallelization at its input for each child.
     ///
-    /// The default implementation returns `true` unless this operator
+    /// If returns `true`, the `ExecutionPlan` would benefit from partitioning
```
Comment (comphead) [inline datafusion/physical-plan/src/lib.rs:191]: 👍

**Review by comphead — COMMENTED** [review]
Thanks @alamb that brings up a lot of explanation on execution plan. I would also love to get a little bit more explanation on what the partition is

File: datafusion/physical-plan/src/lib.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-plan/src/lib.rs:97]: I added a link to `Partitioning` in this PR and I have made another PR to try and document better what partitioning means: https://github.com/apache/arrow-datafusion/pull/8022 -- please take a look when you have time.

**Review by comphead — APPROVED** [review]
LGTM thanks @alamb

**Comment by alamb** [comment]
Thanks again for the review @comphead

**Pull request merged**